builtin/v8/miner/deadline_state.go

```diff
@@ -0,0 +1,1321 @@
+	LiveSectors uint64
+
+	// The total number of sectors in this deadline (incl dead).
+	TotalSectors uint64
+
+	// Memoized sum of faulty power in partitions.
+	FaultyPower PowerPair
+
+	// AMT of optimistically accepted WindowPoSt proofs, submitted during
+	// the current challenge window. At the end of the challenge window,
+	// this AMT will be moved to OptimisticPoStSubmissionsSnapshot. WindowPoSt proofs
+	// verified on-chain do not appear in this AMT.
+	OptimisticPoStSubmissions cid.Cid // AMT[]WindowedPoSt
+
+	// Snapshot of the miner's sectors AMT at the end of the previous challenge
+	// window for this deadline.
+	SectorsSnapshot cid.Cid
+
+	// Snapshot of partition state at the end of the previous challenge
+	// window for this deadline.
+	PartitionsSnapshot cid.Cid
+
+	// Snapshot of the proofs submitted by the end of the previous challenge
+	// window for this deadline.
+	//
+	// These proofs may be disputed via DisputeWindowedPoSt. Successfully
+	// disputed window PoSts are removed from the snapshot.
+	OptimisticPoStSubmissionsSnapshot cid.Cid
+}
+
+type PoStProof struct {
+	PoStProof  abi.RegisteredPoStProof
+	ProofBytes []byte
+}
+
+type WindowedPoSt struct {
+	// Partitions proved by this WindowedPoSt.
+	Partitions bitfield.BitField
+	// Array of proofs, one per distinct registered proof type present in
+	// the sectors being proven. In the usual case of a single proof type,
+	// this array will always have a single element (independent of number
+	// of partitions).
+	Proofs []PoStProof
+}
+
+// Bitwidth of AMTs determined empirically from mutation patterns and projections of mainnet data.
+const DeadlinePartitionsAmtBitwidth = 3 // Usually a small array
+const DeadlineExpirationAmtBitwidth = 5
+
+// Given that 4 partitions can be proven in one post, this AMT's height will
+// only exceed the partition AMT's height at ~0.75EiB of storage.
+const DeadlineOptimisticPoStSubmissionsAmtBitwidth = 2
+
+//
+// Deadlines (plural)
+//
+
+func ConstructDeadlines(emptyDeadlineCid cid.Cid) *Deadlines {
+	d := new(Deadlines)
+	for i := range d.Due {
+		d.Due[i] = emptyDeadlineCid
+	}
+	return d
+}
+
+func (d *Deadlines) LoadDeadline(store adt.Store, dlIdx uint64) (*Deadline, error) {
+	if dlIdx >= uint64(len(d.Due)) {
+		return nil, xc.ErrIllegalArgument.Wrapf("invalid deadline %d", dlIdx)
+	}
+	deadline := new(Deadline)
+	err := store.Get(store.Context(), d.Due[dlIdx], deadline)
+	if err != nil {
+		return nil, xc.ErrIllegalState.Wrapf("failed to lookup deadline %d: %w", dlIdx, err)
+	}
+	return deadline, nil
+}
+
+func (d *Deadlines) ForEach(store adt.Store, cb func(dlIdx uint64, dl *Deadline) error) error {
+	for dlIdx := range d.Due {
+		dl, err := d.LoadDeadline(store, uint64(dlIdx))
+		if err != nil {
+			return err
+		}
+		err = cb(uint64(dlIdx), dl)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (d *Deadlines) UpdateDeadline(store adt.Store, dlIdx uint64, deadline *Deadline) error {
```

Comment:
I don't think including all the mutation methods in this repo is appropriate. It's a very large amount of essentially application logic. There's no way  we'll keep it in sync with the canonical actor code.